### PR TITLE
[SPARK-46710][SQL] Clean up the broadcast data generated when sql execution ends

### PR DESCRIPTION
**Pull request by yabola**
### What changes were proposed in this pull request?
Add config `spark.broadcast.cleanAfterExecution.enabled`(default false) . Clean up the broadcast data generated when sql execution ends ( only suitable for long running Spark SQL services ). 
Before this PR: broadcast data cleaning can only rely on when GC is triggered, which may lead to a lot of waste of memory usage , and may also cause query instability if a single GC takes too long.
After this PR: after the execution of sql is completed, the broadcast data generated during the execution of the sql will be cleared.

Note: this parameter is only suitable for long running Spark SQL services. If this parameter is turned on and one dataframe is collected twice, the broadcast data will not be found during the second execution (because it has been cleaned).

### Why are the changes needed?
 Reduce memory load on driver and executor. This can make a long running spark service more stable. 

### Does this PR introduce _any_ user-facing change?
Add config `spark.broadcast.cleanAfterExecution.enabled`.
Default `false`,  If `true`, the broadcast data generated by the sql will be destroyed when it is completed.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -2025,6 +2025,15 @@ package object config {
       .checkValue(v => v >= 0, "The threshold should be non-negative.")
       .createWithDefault(1L * 1024 * 1024)
 
+    private[spark] val CLEAN_BROADCAST_AFTER_EXECUTION_ENABLED =
+      ConfigBuilder("spark.broadcast.clean_after_execution.enabled")
```
Comment (MaxGekk) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2029]: Please, use the Camel case. Look at the other configs in the file.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 1 above)
Comment (yabola) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Thank you, done

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 2]
```diff
@@ -2025,6 +2025,15 @@ package object config {
       .checkValue(v => v >= 0, "The threshold should be non-negative.")
       .createWithDefault(1L * 1024 * 1024)
 
+    private[spark] val CLEAN_BROADCAST_AFTER_EXECUTION_END_ENABLED =
+      ConfigBuilder("spark.broadcast.cleanAfterExecutionEnd.enabled")
+        .doc("Whether to enable clean broadcast data after sql execution. If enabled, " +
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:2030]: nit:

```suggestion
        .doc("Whether to enable clean broadcast data after SQL execution. If enabled, " +
```

BTW, SQL configurations should go to `SQLConf`. If this is a static conf, it should go to `StaticSQLConf`

Comment (yabola) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: @HyukjinKwon Thank you for your advice. But this parameter needs to be used in `BroadcastManager#initialize` (core module), but the core module does not import the SQL module. The core module use the parameters of the SQL module. I don’t know if they are appropriate.

**Comment by yabola** [comment]
@HyukjinKwon @MaxGekk I have added new UT.  And the failed UT is about `run document build` (I don't know why...)

**Comment by yabola** [comment]
I tested a scenario (a large number of broadcast joins, mainly iceberg expire snapshot & remove_orphan_files SQL), and found its effect to be very noticeable.

**Pull request closed without merging**